assets/js/manual-runner/ManualRunner.tsx

```diff
@@ -0,0 +1,575 @@
+          <div className="flex gap-1 items-center text-base"> <DocumentTextIcon className={`${iconStyle} transition-transform duration-300 group-hover:scale-110 group-hover:text-primary-600`} /> {truncateUid(clip.id)} </div>
+          <div className="text-xs truncate ml-2">{formatDate(new Date(clip.updated_at))}</div>
+        </div>
+      }) :
+        <div className="text-center text-sm">No dataclips found. pick an option above</div>}
+    </div>
+  </>
+}
+
+
+async function readFileContent(file: File): Promise<string> {
+  return new Promise<string>((resolve, reject) => {
+    const reader = new FileReader();
+
+    reader.onload = () => { resolve(reader.result as string); }
+    reader.onerror = () => { reject(reader.error); }
+
+    reader.readAsText(file);
+  });
+}
+
+
+const ImportView: React.FC<{ pushEvent: (event: string, data: any) => void }> = ({ pushEvent }) => {
+  const [importedFiles, setImportedFiles] = React.useState<File[]>([]);
+  const [isValidJSON, setIsValidJSON] = React.useState(true);
+
+  function uploadFiles(f: File[]) {
+    setImportedFiles([...importedFiles, ...f]);
+  }
+
+  function deleteFile(indexImg: number) {
+    setImportedFiles(prev => prev.filter((_, index) => index !== indexImg));
+  }
+
+  React.useEffect(() => {
+    if (importedFiles.length === 1) {
+      const file = importedFiles[0];
+      if (!file) return;
+      void readFileContent(file).then(content => {
+        // check whether JSON is valid
+        try {
+          JSON.parse(content);
+          setIsValidJSON(true);
+          pushEvent("manual_run_change", {
+            manual: {
+              body: content,
+              dataclip_id: null
+            }
+          });
+        } catch (e: any) {
+          setIsValidJSON(false)
+        }
+        return;
+      })
+    } else {
+      setIsValidJSON(true)
+      pushEvent("manual_run_change", {
+        manual: {
+          body: null,
+          dataclip_id: null
+        }
+      });
+    }
+  }, [importedFiles, importedFiles.length, pushEvent])
+
+  return <>
+    {
+      !isValidJSON ?
+        <div className="text-red-700 text-sm flex gap-1 mb-1 items-center"><InformationCircleIcon className={iconStyle} /> File has invalid JSON content</div>
+        :
+        null
+    }
+    <FileUploader
+      currFiles={importedFiles}
+      onUpload={uploadFiles}
+      onDelete={deleteFile}
+      count={1}
+      formats={["json"]} />
+  </>
+}
+
+const EmptyView: React.FC = () => {
+
+  return <div className="flex flex-col items-center gap-2 text-xs py-5">
+    <div className="flex gap-1">
+      <InformationCircleIcon className={`${iconStyle} text-yellow-700`} />
+      An empty input data would be used for this run
+    </div>
+    <div>
+      <div>i.e <span className="bg-slate-700 text-white font-mono p-1 rounded">&#123;&#125;</span></div>
+    </div>
+  </div>
+}
+
+const CustomView: React.FC<{ pushEvent: (event: string, data: any) => void }> = ({ pushEvent }) => {
```

Comment:
These components should all be split out into their own files really

Reply:
Yeah. some functions also need to be in a utils file. additional commits to resolve them.

Reply:
Maybe don't move anything yet, if you can help it, until my edits have been merged down. Let me just check in with Taylor first.

Reply:
merged